tools/filters/sorter.py

```diff
@@ -22,40 +15,48 @@ def stop_err(msg):
 
 def main():
     # define options
-    parser = OptionParser()
-    parser.add_option("-i", "--input")
-    parser.add_option("-o", "--output")
-    parser.add_option("-k", "--key", action="append")
-    parser.add_option("-H", "--header_lines", type='int')
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument("-i", "--input", type=argparse.FileType("r"), help="Tabular file to be sorted")
+    parser.add_argument("-o", "--output", type=argparse.FileType("w"), help="Sorted output file")
+    parser.add_argument("-k", "--key", action="append", help="Key (see manual for bash/sort)")
+    parser.add_argument("-H", "--header_lines", type=int, help="Number of header lines to ignore")
 
     # parse
-    options, args = parser.parse_args()
+    args = parser.parse_args()
 
     try:
         # retrieve options
-        input = options.input
-        output = options.output
-        header_lines = options.header_lines
-        key = [" -k" + k for k in options.key]
-
-        with open(output, 'w') as out:
-            # sed header
-            if header_lines > 0:
-                sed_header = ['sed', '-n', "1,%dp" % header_lines, input]
-                subprocess.check_call(sed_header, stdout=out)
-
-            # grep comments
-            grep_comments = ['grep', '^#', input]
-            exit_code = subprocess.call(grep_comments, stdout=out)
-            if exit_code not in [0, 1]:
-                stop_err('Searching for comment lines failed')
-
-            # grep and sort columns
-            sed_header_restore = ""
-            if header_lines > 0:
-                sed_header_restore = "sed '1,%dd' | " % (header_lines)
-            sort_columns = "(cat %s | %s grep '^[^#]' | sort -f -t '\t' %s)" % (input, sed_header_restore, ' '.join(key))
-            subprocess.check_call(sort_columns, stdout=out, shell=True)
+        input_fh = args.input
+        output_fh = args.output
+        header_lines = args.header_lines
+        key_args = []
+        for k in args.key:
+            key_args.extend(['-k', k])
+
+        # sed header
+        if header_lines > 0:
+            sed_header = ['sed', '-n', f"1,{header_lines:d}p"]
+            subprocess.check_call(sed_header, stdin=input_fh, stdout=output_fh)
+            input_fh.seek(0)
+
+        # grep comments
+        grep_comments = ['grep', '^#']
```

Comment:
I'm wondering how this works? This grep does not get any input, or am I wrong. `planemo test` stalls here as I would expect, but if I try on usegalaxy.eu it runs through.

Will work on this over here https://github.com/galaxyproject/galaxy/pull/15229

> This is a refactoring of components with existing test coverage.

Turns out it isn't .. but I also always assumed that built in tools are tested.

Reply:
Yep, looks like a bug, the original version had `input` at the end of the list (now renamed `input_fh`).

Reply:
Hi guys, this is still broken as I got hit by it in a recent Galaxy deployment.

This script just hangs and blocks up handlers if used by users, as it greps empty space.

Can it be patched?